src/pretty/chain.rs

```diff
@@ -0,0 +1,237 @@
+        for node in nodes {
+            if operand_pred(node) {
+                self.chain_len += 1;
+                let mut seen_op = false;
+                for child in node.children() {
+                    if let Some(op) = op_converter(child) {
+                        seen_op = true;
+                        self.items.push(ChainItem::Op(op));
+                    } else if is_comment_node(child) {
+                        let doc = self.printer.convert_comment(child);
+                        self.items.push(if can_attach {
+                            ChainItem::Attached(doc)
+                        } else {
+                            ChainItem::Comment(doc)
+                        });
+                        self.has_comment = true;
+                    } else if child.kind() == SyntaxKind::Space {
+                        if child.text().has_linebreak() {
+                            if self.items.last().is_some_and(|last| {
+                                matches!(*last, ChainItem::Attached(_) | ChainItem::Comment(_))
+                            }) {
+                                self.items.push(ChainItem::Linebreak);
+                            }
+                            can_attach = false;
+                        }
+                    } else if seen_op {
+                        if let Some(rhs) = rhs_converter(child) {
+                            self.items.push(ChainItem::Body(rhs));
+                            can_attach = true;
+                        }
+                    }
+                }
+            } else if let Some(fallback) = fallback_converter(node) {
+                // We must use this to handle args.
+                if let Some(ChainItem::Body(body)) = self.items.last_mut() {
+                    *body += fallback;
+                } else {
+                    self.items.push(ChainItem::Body(fallback));
+                }
+            }
+        }
+
+        // for item in &self.items {
+        //     println!(
+        //         "+{}",
+        //         match item {
+        //             ChainItem::Body(doc) => format!("Body {:?}", doc.pretty(120).to_string()),
+        //             ChainItem::Op(doc) => format!("Op {:?}", doc.pretty(120).to_string()),
+        //             ChainItem::Comment(doc) => format!("Comment {:?}", doc.pretty(120).to_string()),
+        //             ChainItem::Attached(doc) =>
+        //                 format!("Attached {:?}", doc.pretty(120).to_string()),
+        //             ChainItem::Linebreak => format!("Linebreak"),
+        //         }
+        //     );
+        // }
+
+        self
+    }
+
+    pub fn print_doc(self, sty: ChainStyle) -> ArenaDoc<'a> {
+        let arena = &self.printer.arena;
+
+        let op_sep = if sty.space_around_op {
+            arena.line()
+        } else {
+            arena.line_()
+        };
+
+        let use_simple_layout = self.chain_len == 1 && sty.no_break_single && !self.has_comment;
```

Comment:
I suppose that `chain_len` means the number of operators in chain?

Reply:
yes. perhaps `op_num` is a better name with less ambiguity.